### PR TITLE
Use frozen strings by default


### PR DESCRIPTION
**Pull request by p-mongo**

### CODE REVIEW

File: lib/mongo/error/parser.rb [hunk 1]
```diff
@@ -148,7 +151,7 @@ def write_concern_error_labels
       private
 
       def parse!
-        @message = ""
+        @message = +""
```
Comment (alexbevi) [inline lib/mongo/error/parser.rb:154]: typo?

File: lib/mongo/grid/file/chunk.rb [hunk 2]
```diff
@@ -153,7 +156,7 @@ class << self
           # @since 2.0.0
           # @api private
           def assemble(chunks)
-            chunks.reduce(''){ |data, chunk| data << chunk.data.data }
+            chunks.reduce(+''){ |data, chunk| data << chunk.data.data }
```
Comment (alexbevi) [inline lib/mongo/grid/file/chunk.rb:159]: typo?

File: lib/mongo/server/connection_common.rb [hunk 3]
```diff
@@ -88,7 +91,7 @@ def add_server_diagnostics
         # Server::Monitor::Connection does not reference its server, but
         # knows its address. Server::Connection delegates the address to its
         # server.
-        note = "on #{address.seed}"
+        note = +"on #{address.seed}"
```
Comment (alexbevi) [inline lib/mongo/server/connection_common.rb:94]: typo

File: spec/integration/bson_symbol_spec.rb [hunk 4]
```diff
@@ -12,7 +15,7 @@
   end
 
   let(:expected) do
-    "\x12\x00\x00\x00\x0Efoo\x00\x04\x00\x00\x00foo\x00\x00".force_encoding('binary')
+    (+"\x12\x00\x00\x00\x0Efoo\x00\x04\x00\x00\x00foo\x00\x00").force_encoding('binary')
```
Comment (alexbevi) [inline spec/integration/bson_symbol_spec.rb:18]: typo?

File: spec/integration/grid_fs_bucket_spec.rb [hunk 5]
```diff
@@ -30,7 +33,7 @@
 
     context 'in binary encoding' do
       let(:data_to_write) do
-        data.force_encoding('binary').freeze
+        data.dup.force_encoding('binary').freeze
```
Comment (alexbevi) [inline spec/integration/grid_fs_bucket_spec.rb:36]: this should likely be in a standalone commit so the purpose of the change is clearer (it's not just adding magic comments)

File: spec/mongo/grid/file/chunk_spec.rb [hunk 6]
```diff
@@ -57,7 +60,7 @@
     end
 
     let(:raw_data) do
-      'testing'
+      +'testing'
```
Comment (alexbevi) [inline spec/mongo/grid/file/chunk_spec.rb:63]: typo

File: spec/mongo/grid/file/chunk_spec.rb [hunk 7]
```diff
@@ -150,7 +153,7 @@
     context 'when the data is smaller than the default size' do
 
       let(:raw_data) do
-        'testing'
+        +'testing'
```
Comment (alexbevi) [inline spec/mongo/grid/file/chunk_spec.rb:156]: typo

File: spec/mongo/grid/file/chunk_spec.rb [hunk 8]
```diff
@@ -185,15 +188,15 @@
       end
 
       let(:raw_data) do
-        'testing'
+        +'testing'
```
Comment (alexbevi) [inline spec/mongo/grid/file/chunk_spec.rb:191]: typo

File: spec/mongo/grid/file/chunk_spec.rb [hunk 9]
```diff
@@ -185,15 +188,15 @@
...
       end
 
       let(:data) do
         BSON::Binary.new(raw_data)
       end
 
       let(:assembled) do
-        full_data = ''
+        full_data = +''
```
Comment (alexbevi) [inline spec/mongo/grid/file/chunk_spec.rb:199]: typo

File: spec/mongo/grid/file_spec.rb [hunk 10]
```diff
@@ -44,7 +47,7 @@
     end
 
     let(:data) do
-      'testing'
+      +'testing'
```
Comment (alexbevi) [inline spec/mongo/grid/file_spec.rb:50]: typo

File: spec/support/matchers.rb [hunk 11]
```diff
@@ -20,7 +23,7 @@
 
 RSpec::Matchers.define :be_cstring do |string|
   match do |actual|
-    actual == "#{string.force_encoding(BSON::BINARY)}\0"
+    actual == "#{string.dup.force_encoding(BSON::BINARY)}\0"
```
Comment (alexbevi) [inline spec/support/matchers.rb:26]: Should this be in a separate commit to differentiate it from this PR?

**Review by alexbevi — CHANGES_REQUESTED** [review]

File: spec/support/utils.rb [hunk 12]
```diff
@@ -330,7 +333,7 @@ def int64_value(value)
     else
       ''
     end
-    uri = "mongodb://#{creds}#{address_strs.join(',')}/"
+    uri = +"mongodb://#{creds}#{address_strs.join(',')}/"
```
[the remark this answers is not in the document]
Comment (p-mongo) [reply spec/support/utils.rb:336]: No, unary plus makes the string mutable. https://stackoverflow.com/questions/37799296/ruby-what-does-the-comment-frozen-string-literal-true-do

File: lib/mongo/auth/stringprep/unicode_normalize/normalize.rb [hunk 13]
```diff
@@ -1,3 +1,6 @@
+# frozen_string_literal: true
+# encoding: utf-8
+
```
[the remark this answers is not in the document]
Comment (p-mongo) [reply lib/mongo/auth/stringprep/unicode_normalize/normalize.rb:3]: Fixed these.

File: spec/support/matchers.rb (hunk 11 above)
Comment (p-mongo) [reply spec/support/matchers.rb:26]: I do not understand what you mean. This PR makes string literals produce frozen strings. Sometimes the strings thus produced are modified by the code, in which case they need to be unfrozen. This is such an occurrence.

File: spec/support/utils.rb (hunk 12 above)
[the remark this answers is not in the document]
Comment (alexbevi) [reply spec/support/utils.rb:336]: Wow, TIL. Thanks!

File: spec/support/matchers.rb (hunk 11 above)
Comment (alexbevi) [reply spec/support/matchers.rb:26]: Yeah this was my mistake. I should have done a bit more research here.

**Review by durran — APPROVED** [review]
TIL about those magic comments and `+` on string.

**Review by alexbevi — APPROVED** [review]
